### ISSUE TITLE
Detection

### ISSUE THREAD

**Issue by Y-J-Q-L** [roymiles/Simple-Recipe-Distillation#2]
Thank you very much for your work. Would it be possible for you to provide the implementation code for object detection?

**Comment by roymiles** [comment]
Hi @Y-J-Q-L unfortunately I don't think I have the code for this experiment anymore. However, I could recommend using the code we released [here](https://github.com/roymiles/VkD/tree/master/vidt) on COCO 2017 but with the Swin architecture (rather than YOLO).  Also [here](https://github.com/roymiles/VkD/blob/6f8a5072447a1c5bb6043cdc035cf7b78d3854d8/vidt/methods/vidt/criterion.py#L298) is where the loss is implemented.

**Comment by Jocelyn-Jiaxin** [comment]
Thank you very much for your reply.
 I would like to cite your method in my paper for comparative experiments, so I apologize for bothering you again. Could you please remind me of the composition of the distillation loss function used in your object detection (Faster R-CNN) approach? Did it only utilize (https://github.com/roymiles/Simple-Recipe-Distillation/blob/main/imagenet/torchdistill/losses/single.py#L150)?

**Comment by roymiles** [comment]
Ah right, so for this paper we did use both the distillation loss function and the original object detection task losses (equal weighting). We also performed the normalisation (i.e. the subtracting mean and dividing by standard deviation) over the height * width dimensions as opposed to the batch dimension.

Hope that help! let me know if there any other questions and I can try and be more specific.

**Comment by Jocelyn-Jiaxin** [comment]
Thank you so much for your response! I was able to successfully replicate your code. I really appreciate your patience and help, and I wish you continued success in your research.

**Comment by roymiles** [comment]
That's great news! No worries at all and I wish you all the best with your research too! 😊 always interested to read more about others working in these directions.

**Comment by Jocelyn-Jiaxin** [comment]
Hello, I apologize for bothering you again. I have implemented the code as follows, but I found that when I apply normalization (as shown in the commented section) in the object detection task, my loss becomes negative. After carefully reviewing the code, I would like to ask whether it is still necessary to use BatchNorm1d in the object detection task.
@MODELS.register_module()
class SGDLoss(nn.Module):

    def __init__(self, reduction='mean', loss_weight=1.0):
        super(SGDLoss, self).__init__()
        self.reduction = reduction
        self.loss_weight = loss_weight
        t_dim = 256 
        s_dim = 256 
        self.embed = nn.Linear(s_dim, t_dim).cuda()
        self.bn_s = torch.nn.BatchNorm1d(t_dim, eps=0.0001, affine=False).cuda()
        self.bn_t = torch.nn.BatchNorm1d(t_dim, eps=0.0001, affine=False).cuda()
        # self.gn_s = torch.nn.GroupNorm(16, t_dim, eps=0.0, affine=False).cuda()
        # self.gn_t = torch.nn.GroupNorm(16, t_dim, eps=0.0, affine=False).cuda()

    def forward(self,
                pred,
                target,
                weight=None,
                avg_factor=None,
                reduction_override=None) -> torch.Tensor:
        assert reduction_override in (None, 'none', 'mean', 'sum')
        reduction = (
            reduction_override if reduction_override else self.reduction)
        # pred_mean = pred.mean(dim=(2, 3), keepdim=True)
        # pred_std = pred.std(dim=(2, 3), keepdim=True)
        # target_mean = target.mean(dim=(2, 3), keepdim=True)
        # target_std = target.std(dim=(2, 3), keepdim=True)
        # pred =(pred - pred_mean) / (pred_std + 1e-6)
        # target = (target - target_mean) / (target_std + 1e-6)

        z_s = pred.mean(-1).mean(-1)  # Bx256
        z_t = target.mean(-1).mean(-1)  # Bx256

        # Rep loss
        loss = 0.1 * self.forward_loss(z_s, z_t)
        return loss
    def forward_loss(self, z_s, z_t):
        f_t = z_t

        f_s = self.embed(z_s)
        n, d = f_s.shape

        f_s_norm = self.bn_s(f_s)
        f_t_norm = self.bn_t(f_t) 

        c_diff = f_s_norm - f_t_norm
        c_diff = torch.abs(c_diff) 
        c_diff = c_diff.pow(5.0)

        loss = torch.log(c_diff.sum())
        return loss

**Comment by roymiles** [comment]
Hi, no worries at all!

The commented code looks right for standardising over the H*W (i.e. BN over H*W). A negative loss here is not strictly bad, as it simply means that the argument inside the log is < 1. You can artifically avoid this going negative by doing `torch.log(c_diff.sum() + 1.)`. 

Is the model still training and learning (ignoring the negative distillation loss)? A negative value should just mean that the absolute difference is very small! (which is good). Furthermore, the batchnorm might not be strictly necessary, but in our experiments we found that using it over the H*W-dims (like you do in the commented code) did improve the results over not using it.

**Comment by Jocelyn-Jiaxin** [comment]
Hello, I apologize for bothering you again. I think I've found the reason why the feature values are negative. If we normalize pred across the H×W dimensions and then perform z_s = pred.mean(-1).mean(-1), it will result in z_s being all zeros, which makes distillation impossible. Do you happen to recall what the specific loss calculation was like here?
Perhaps I didn't clarify this earlier, but I'm employing ResNet101 as the teacher model's backbone and ResNet18 as the student model's backbone. I'm uncertain whether this configuration would necessitate any modifications to the loss function.
Below is the code for my loss function. In this context, pred denotes the feature maps extracted from the student model's neck layer, while target denotes the feature maps extracted from the teacher model's neck layer.
`class SGDLoss(nn.Module):

    def __init__(self, reduction='mean', loss_weight=1.0):
        super(SGDLoss, self).__init__()
        self.reduction = reduction
        self.loss_weight = loss_weight
        t_dim = 256 
        s_dim = 256 
        self.embed = nn.Linear(s_dim, t_dim).cuda()
        self.bn_s = torch.nn.BatchNorm1d(t_dim, eps=0.0001, affine=False).cuda()
        self.bn_t = torch.nn.BatchNorm1d(t_dim, eps=0.0001, affine=False).cuda()
        # self.gn_s = torch.nn.GroupNorm(16, t_dim, eps=0.0, affine=False).cuda()
        # self.gn_t = torch.nn.GroupNorm(16, t_dim, eps=0.0, affine=False).cuda()

    def forward(self,
                pred,
                target,
                weight=None,
                avg_factor=None,
                reduction_override=None) -> torch.Tensor:
        assert reduction_override in (None, 'none', 'mean', 'sum')
        reduction = (
            reduction_override if reduction_override else self.reduction)
        pred_mean = pred.mean(dim=(2, 3), keepdim=True)
        pred_std = pred.std(dim=(2, 3), keepdim=True)
        target_mean = target.mean(dim=(2, 3), keepdim=True)
        target_std = target.std(dim=(2, 3), keepdim=True)
        pred =(pred - pred_mean) / (pred_std)
        target = (target - target_mean) / (target_std)

        z_s = pred.mean(-1).mean(-1)  # Bx256
        z_t = target.mean(-1).mean(-1)  # Bx256

        # Rep loss
        loss = 0.1 * self.forward_loss(z_s, z_t)
        return loss
    def forward_loss(self, z_s, z_t):
        f_t = z_t

        f_s = self.embed(z_s)
        n, d = f_s.shape

        f_s_norm = self.bn_s(f_s)
        f_t_norm = self.bn_t(f_t) 

        c_diff = f_s_norm - f_t_norm
        c_diff = torch.abs(c_diff) 
        c_diff = c_diff.pow(5.0)

        loss = torch.log(c_diff.sum()+1.)
        return loss`

**Comment by roymiles** [comment]
Ah yes, sorry I overlooked that bit. You are right, you don't sum over the H*W-dims if you normalise them to have zero mean.
Can you try removing the mean and batch norm ops? like:

```
def __init__(self, reduction='mean', loss_weight=1.0):
    super(SGDLoss, self).__init__()
    self.reduction = reduction
    self.loss_weight = loss_weight
    t_dim = 256 
    s_dim = 256 
    self.embed = nn.Linear(s_dim, t_dim).cuda()

def forward(self,
            pred,
            target,
            weight=None,
            avg_factor=None,
            reduction_override=None) -> torch.Tensor:
    assert reduction_override in (None, 'none', 'mean', 'sum')
    reduction = (reduction_override if reduction_override else self.reduction)
    pred_mean = pred.mean(dim=(2, 3), keepdim=True)
    pred_std = pred.std(dim=(2, 3), keepdim=True)
    target_mean = target.mean(dim=(2, 3), keepdim=True)
    target_std = target.std(dim=(2, 3), keepdim=True)
    pred =(pred - pred_mean) / (pred_std)
    target = (target - target_mean) / (target_std)

    # b, c, h, w -> b, h, w, c - self.embed is over last dim
    z_s = torch.einsum('bchw->bhwc', pred)
    z_t = torch.einsum('bchw->bhwc', target)

    # Rep loss
    loss = 0.1 * self.forward_loss(z_s, z_t)
    return loss
def forward_loss(self, z_s, z_t):
    f_t = z_t

    f_s = self.embed(z_s)
    n, d = f_s.shape

    c_diff = f_s - f_t
    c_diff = torch.abs(c_diff) 
    c_diff = c_diff.pow(5.0)

    loss = torch.log(c_diff.sum())
    return loss
```

Taking the features extracted from the neck layer makes sense. I think it is quite empirical, but I found choosing features with the smallest H*W dim and largest C-dim works best, while being as deep as possible. I can check through some of my other code when I'm back next week.

**Comment by Jocelyn-Jiaxin** [comment]
Hello, I tried out your code. However, I noticed that there was very little improvement in the loss, and the distillation results remain far from ideal.
2024/12/27 11:38:55 - mmengine - INFO - Epoch(train)  [1][ 50/656]  lr: 9.9098e-04  eta: 0:50:41  time: 0.3889  data_time: 0.0381  memory: 7865  loss: 5.5009  loss_cls: 0.2076  loss_bbox: 1.5914  loss_dfl: 0.6344  loss_cls_kd: 0.0000  loss_reg_kd: 0.0000  loss_feat_kd: 3.0675
2024/12/27 11:39:09 - mmengine - INFO - Epoch(train)  [1][100/656]  lr: 1.9920e-03  eta: 0:43:20  time: 0.2805  data_time: 0.0106  memory: 5190  loss: 4.6885  loss_cls: 0.2939  loss_bbox: 1.0798  loss_dfl: 0.4188  loss_cls_kd: 0.0000  loss_reg_kd: 0.0000  loss_feat_kd: 2.8960
2024/12/27 11:39:30 - mmengine - INFO - Epoch(train)  [1][150/656]  lr: 2.9930e-03  eta: 0:46:25  time: 0.4130  data_time: 0.0075  memory: 4907  loss: 4.7658  loss_cls: 0.3831  loss_bbox: 1.0451  loss_dfl: 0.4597  loss_cls_kd: 0.0000  loss_reg_kd: 0.0000  loss_feat_kd: 2.8779
2024/12/27 11:39:52 - mmengine - INFO - Epoch(train)  [1][200/656]  lr: 3.9940e-03  eta: 0:48:44  time: 0.4426  data_time: 0.0076  memory: 5273  loss: 4.8807  loss_cls: 0.5841  loss_bbox: 1.0189  loss_dfl: 0.4024  loss_cls_kd: 0.0000  loss_reg_kd: 0.0000  loss_feat_kd: 2.8753
2024/12/27 11:40:14 - mmengine - INFO - Epoch(train)  [1][250/656]  lr: 4.9950e-03  eta: 0:49:58  time: 0.4419  data_time: 0.0072  memory: 4908  loss: 4.6156  loss_cls: 0.4968  loss_bbox: 0.8791  loss_dfl: 0.3672  loss_cls_kd: 0.0000  loss_reg_kd: 0.0000  loss_feat_kd: 2.8726
2024/12/27 11:40:36 - mmengine - INFO - Epoch(train)  [1][300/656]  lr: 5.9960e-03  eta: 0:50:34  time: 0.4380  data_time: 0.0074  memory: 4906  loss: 4.3761  loss_cls: 0.5278  loss_bbox: 0.6791  loss_dfl: 0.3100  loss_cls_kd: 0.0000  loss_reg_kd: 0.0000  loss_feat_kd: 2.8592
2024/12/27 11:40:59 - mmengine - INFO - Epoch(train)  [1][350/656]  lr: 6.9970e-03  eta: 0:51:14  time: 0.4567  data_time: 0.0077  memory: 6119  loss: 4.6473  loss_cls: 0.5361  loss_bbox: 0.8917  loss_dfl: 0.3674  loss_cls_kd: 0.0000  loss_reg_kd: 0.0000  loss_feat_kd: 2.8522
2024/12/27 11:41:20 - mmengine - INFO - Epoch(train)  [1][400/656]  lr: 7.9980e-03  eta: 0:51:17  time: 0.4337  data_time: 0.0080  memory: 5768  loss: 4.3445  loss_cls: 0.3869  loss_bbox: 0.7733  loss_dfl: 0.3388  loss_cls_kd: 0.0000  loss_reg_kd: 0.0000  loss_feat_kd: 2.8454
2024/12/27 11:41:42 - mmengine - INFO - Epoch(train)  [1][450/656]  lr: 8.9990e-03  eta: 0:51:15  time: 0.4339  data_time: 0.0071  memory: 4915  loss: 4.4922  loss_cls: 0.4705  loss_bbox: 0.8275  loss_dfl: 0.3488  loss_cls_kd: 0.0000  loss_reg_kd: 0.0000  loss_feat_kd: 2.8454
2024/12/27 11:42:04 - mmengine - INFO - Epoch(train)  [1][500/656]  lr: 1.0000e-02  eta: 0:51:18  time: 0.4467  data_time: 0.0084  memory: 4698  loss: 4.4327  loss_cls: 0.4994  loss_bbox: 0.7627  loss_dfl: 0.3286  loss_cls_kd: 0.0000  loss_reg_kd: 0.0000  loss_feat_kd: 2.8420
2024/12/27 11:42:26 - mmengine - INFO - Epoch(train)  [1][550/656]  lr: 1.0000e-02  eta: 0:51:12  time: 0.4395  data_time: 0.0091  memory: 4907  loss: 4.6155  loss_cls: 0.7809  loss_bbox: 0.6807  loss_dfl: 0.3130  loss_cls_kd: 0.0000  loss_reg_kd: 0.0000  loss_feat_kd: 2.8409
2024/12/27 11:42:49 - mmengine - INFO - Epoch(train)  [1][600/656]  lr: 1.0000e-02  eta: 0:51:05  time: 0.4430  data_time: 0.0078  memory: 6815  loss: 4.3700  loss_cls: 0.5770  loss_bbox: 0.6674  loss_dfl: 0.2876  loss_cls_kd: 0.0000  loss_reg_kd: 0.0000  loss_feat_kd: 2.8379
2024/12/27 11:43:10 - mmengine - INFO - Epoch(train)  [1][650/656]  lr: 1.0000e-02  eta: 0:50:52  time: 0.4364  data_time: 0.0085  memory: 5598  loss: 4.2686  loss_cls: 0.5158  loss_bbox: 0.6446  loss_dfl: 0.2768  loss_cls_kd: 0.0000  loss_reg_kd: 0.0000  loss_feat_kd: 2.8315

**Comment by roymiles** [comment]
Hi, it is hard for me to make any conclusions from this log over half an epoch. I am also not sure what the losses of the baseline experiment look like. How do the training losses change over the entire run? as far as I can see, the losses are going down and the exact values are not particularly important, more so the final evaluation accuracy?

As a comparison, I have another set of logs for a similar(ish) object detection distillation run [here](https://drive.google.com/drive/folders/1QvUDcI2Va6QLK5jxdL6DAur1-jlWuPe_ ):log.txt:train_loss_distil. The distillation loss is averaged after every epoch, with values: 6.868880860879362, 5.439398912717214, 4.965710392868053, 4.7177099857071685, ...
However, this is without a `log()` loss so a drop in 1.4 is less significant than the 0.2 drop you observe in your training log.

**Comment by Jocelyn-Jiaxin** [comment]
Thank you very much for your reply. I accidentally overlooked this email and just saw it now. Yes, you are absolutely right, I neglected the impact of the logarithm on the change in loss.

Thank you again for your patient response. I will cite and compare your method in my paper. I wish you all the best in your research.
